### PR TITLE
HAWQ-969. Add getting configuration from HDFS and YARN

### PR DESCRIPTION
**Pull request by wcl14**
Add getting configuration from HDFS and YARN, and writing xml file in xml_parser.cpp.

### CODE REVIEW

File: src/test/feature/lib/hdfs_config.h [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+     * set parameter value in ./etc/hadoop/hdfs-site.xml according to parameter name
+     * @param parameterName, parameter name which used to set parameter value
+     * @param parameterValue, parameter value which to be set
+     * @return true if succeeded
+     */
+    bool setParameterValue(const std::string &parameterName, const std::string &parameterValue);
+
+  private:
+    /**
+     * @return yarn user
```
Comment (xunzhang) [inline src/test/feature/lib/hdfs_config.h]: TYPO: `return hdfs user`. Actually, I do not suggest that we should add doxygen-style comments of private interfaces.

File: src/test/feature/lib/hdfs_config.h [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+ */
+class HdfsConfig {
+  public:
+    /**
+     * HdfsConfig constructor
+     */
+    HdfsConfig(): psql(HAWQ_DB, HAWQ_HOST, HAWQ_PORT, HAWQ_USER, HAWQ_PASSWORD) {}
+
+    /**
+     * HdfsConfig desstructor
```
Comment (xunzhang) [inline src/test/feature/lib/hdfs_config.h]: ...typo

File: src/test/feature/lib/hdfs_config.cpp [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+  string confPath=getHadoopHome();
+  if (confPath == "")
+    return false;
+  confPath.append("/etc/hadoop/hdfs-site.xml");
+  hdfsxmlconf.reset(new XmlConfig(confPath));
+  hdfsxmlconf->parse();
+  return true;
+}
+
+bool HdfsConfig::isHA() {
```
Comment (xunzhang) [inline src/test/feature/lib/hdfs_config.cpp]: [REVIEW-LINK-1] I think the implementation of these functions are problematic. Say if user modified these configuration files without restart HAWQ, the result could be confused. A better way is to get these kinds of status from running HAWQ, is that possible?

File: src/test/feature/testlib/test_lib.cpp [hunk 4]
```diff
@@ -56,6 +58,69 @@ TEST_F(TestCommonLib, TestHawqConfig) {
   return;
 }
 
+TEST_F(TestCommonLib, TestHdfsConfig) {
+  hawq::test::HdfsConfig hc;
+  hc.isHA();
```
Comment (xunzhang) [inline src/test/feature/testlib/test_lib.cpp]: As I mentioned in **[REVIEW-LINK-1]**, if you implemented with loading from a configuration file, you should not just invoke the interface. You should load from a test config file and use `EXPECT_EQ`-like macro function to write your tests.

File: src/test/feature/lib/hdfs_config.cpp [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+bool HdfsConfig::isTruncate() {
+  string cmd = "hadoop fs -truncate";
+  Command c(cmd);
+  string result = c.run().getResultOutput();
+  auto lines = hawq::test::split(result, '\n');
+  if (lines.size() >= 1) {
+    string valueLine = lines[0];
+    int find = valueLine.find("-truncate: Unknown command");
+    if (find < 0)
```
Comment (xunzhang) [inline src/test/feature/lib/hdfs_config.cpp]: Unify your coding style with parentheses

File: src/test/feature/lib/hdfs_config.cpp [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+  }
+
+  hawqxmlconf.reset(new XmlConfig(confPath));
+  hawqxmlconf->parse();
+  return true;
+}
+
+bool HdfsConfig::LoadFromHdfsConfigFile() {
+  string confPath=getHadoopHome();
+  if (confPath == "")
```
Comment (xunzhang) [inline src/test/feature/lib/hdfs_config.cpp]: Unify your coding style with parentheses

File: src/test/feature/lib/hdfs_config.cpp [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool HdfsConfig::isKerbos() {
+  bool ret = LoadFromHawqConfigFile();
+  if (!ret) {
+    return false;
```
Comment (xunzhang) [inline src/test/feature/lib/hdfs_config.cpp]: What is the semantics of the interface returning `false`? If it means that "without Kerbos", the implementation is incorrect. You should throw exception or load default configuration instead.

File: src/test/feature/lib/hdfs_config.h [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+     * whether HDFS is in HA mode
+     * @return true if HDFS is HA
+     */
+    bool isHA();
+
+    /**
+     * whether HDFS is kerbos
+     * @return true if HDFS is kerbos
+     */
+    bool isKerbos();
```
Comment (xunzhang) [inline src/test/feature/lib/hdfs_config.h]: I am not sure kerbos is a short name of kerberos. I have googled that, I think you use a incorrect library name.

File: src/test/feature/lib/hdfs_config.h [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+     * @return true if getHANamenode succeeded
+     */
+    bool getHANamenode(const std::string &namenodetype,
+                       std::string &namenode,
+                                   int &port);
+
+  private:
+    std::unique_ptr<XmlConfig> hawqxmlconf;
+    std::unique_ptr<XmlConfig> hdfsxmlconf;
+    hawq::test::PSQL psql;
```
Comment (xunzhang) [inline src/test/feature/lib/hdfs_config.h]: You don't need psql object in this class. Remember modifying the construction too. Same problem also exists in `YarnConfig`.

File: src/test/feature/lib/yarn_config.cpp [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+      port.push_back(std::stoi(hawq::test::trim(hawq::test::split(values, ':')[1])));
+    }
+  }
+}
+
+
+string YarnConfig::getParameterValue(const string &parameterName) {
+  bool ret = LoadFromYarnConfigFile();
+  if (!ret) {
+    return NULL;
```
Comment (xunzhang) [inline src/test/feature/lib/yarn_config.cpp]: return type is `string`...

File: src/test/feature/lib/xml_parser.h [hunk 11]
```diff
@@ -17,10 +17,27 @@ class XmlConfig {
...
+  void closeAndSave();
+
   // parse the configuration file
   void parse();
 
   // @param key The key of the configuration item
+  // @param value The updated value
+  // @param save whether save the updated document to disk, if save is false, open() and closeAndSave() should be called additionally
+  // @ return The value of configuration item
+  bool setString(const std::string &key, const std::string &value, const bool &save);
```
Comment (xunzhang) [inline src/test/feature/lib/xml_parser.h]: With POD types, there is no case using const reference. For example, here we should define as `bool setString(const std::string &key, const std::string &value, bool save);`

**Comment by xunzhang** [comment]
LGTM

File: src/test/feature/lib/hdfs_config.cpp [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+  const char *env = getenv("GPHOME");
+  string confPath = env ? env : "";
+  if (confPath != "") {
+    confPath.append("/etc/hdfs-client.xml");
+  } else {
+    return false;
+  }
+
+  hawqxmlconf.reset(new XmlConfig(confPath));
+  hawqxmlconf->parse();
```
Comment (wengyanqing) [inline src/test/feature/lib/hdfs_config.cpp]: Your should handle parse fail condition.

File: src/test/feature/lib/hdfs_config.cpp [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+    int find = valueLine.find("-truncate: Unknown command");
+    if (find < 0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+string HdfsConfig::getHadoopHome() {
+  string cmd = "ps -ef|grep hadoop";
```
Comment (wengyanqing) [inline src/test/feature/lib/hdfs_config.cpp]: I see lots of places have the same logic. 
1. run command
2. get command result by lines
3. process lines result
It's better to extract the common logic into function.

File: src/test/feature/lib/hdfs_config.cpp [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+  if (!ret) {
+    throw GetHadoopHomeException();
+  }
+
+  return hdfsxmlconf->getString(parameterName);
+}
+
+string HdfsConfig::getParameterValue(const string &parameterName,
+                                     const string &conftype) {
+  if (conftype == "hdfs" || conftype == "HDFS")
```
Comment (wengyanqing) [inline src/test/feature/lib/hdfs_config.cpp]: It's better to call lower(conftype) == "hdfs" ?

File: src/test/feature/lib/hdfs_config.h [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+namespace hawq {
+namespace test {
+
+/**
+ * HdfsConfig common libray. Get detailed information about HDFS
+ * including checking state of namenodes and datanodes, get parameter value
+ * @author Chunling Wang
+ */
+class HdfsConfig {
```
Comment (wengyanqing) [inline src/test/feature/lib/hdfs_config.h]: Should HdfsConfig be singleton ?

**Comment by wengyanqing** [comment]
LGTM

File: src/test/feature/lib/hdfs_config.cpp (hunk 14 above)
Comment (xunzhang) [inline src/test/feature/lib/hdfs_config.cpp]: `hawq::test::lower(string)`

**Pull request closed without merging**